GH-3891 - Configure quarantineMode

rob4629 opened the pull request:
<!--
Thank you for your contribution.

Before making a PR, please read our contributing guidelines at
https://github.com/DevExpress/testcafe/blob/master/CONTRIBUTING.md#code-contribution

We recommend creating a *draft* PR, so that you can mark it as 'ready for review' when you are done.
-->

## Purpose
_Describe the problem you want to address or the feature you want to implement._
Closes #3891 : Allow for Quarantine threshold values to be configured (_instead of hardcoded to 3_). This should include the number of retries, as well as the minimum number of passes.

## Approach
_Describe how your changes address the issue or implement the desired functionality in as much detail as possible._
I amended `quarantineMode` to optionally accept a Dictionary. Parse the input (and validate), then use this to set the threshold values of `failedThresholdReached` and `passedThresholdReached`

## References
_Provide a link to the existing issue(s), if any._
#3891 

## Pre-Merge TODO
- [x] Write tests for your proposed changes
- [ ] Make sure that existing tests do not fail

rob4629 commented on src/runner/test-run-controller.ts:
```diff
@@ -199,6 +211,12 @@ export default class TestRunController extends AsyncEventEmitter {
     private async _emitTestRunStart (): Promise<void> {
         await this.emit('test-run-start');
+        if (this._quarantine) {
+            // const { passCount, retryCount } = this._opts.quarantineMode;
+
+            this._quarantine.setFailedQuarantineThreshold(3);
+            this._quarantine.setPassedQuarantineThreshold(1);
```
Temporarily hardcoded, while I figure out why `this._opts.quarantineMode` is still just a boolean (instead of including `passCount` and `retryCount` values)

miherlosev commented:
Hi @rob4629
Thank you for contributing to TestCafe. You expose internal `failedThreshold` and `passedThreshold` parameters via `quarantineMode { "passCount": 1, "retryCount": 5 }` API.
Description of parameters:
* `passCount` is the number of passed test runs to consider the test passed
* `retryCount` is the number of failed test runs to consider the test failed

In general, this approach is good.
At the same time, we believe that the `retryCount` parameter should be a full test run count that TestCafe will execute.
For example:
* consider the test passed if at least one run passed - `quarantineMode { "passCount": 1 }`
* increase the number of test run attempts - `quarantineMode { "retryCount": 10 }

This approach doesn't seem to require a lot of changes. You need to calculate the `failedThreshold` value as a result of the `retryCount` minus `passCount` expression and validate if that `passCount` is less than `retryCount`.

rob4629 commented:
Thanks for the feedback @miherlosev. I can implement that change today. Does `passCount` seem a sensible/explicit name to you, or should this be `passThreshold`?

Also, it looks like my change has broken existing functionality:
```sh
testcafe -q tests/myFixture/myTest.js

ERROR The "--quarantine-mode" option value is not a valid key-value pair.
```

If I explicitly add the testpath argument afterwards, it behaves as expected
```sh
testcafe -q -t tests/myFixture/myTest.js
```

Although I'm validating it's a valid Key-Value pair in `getQuarantineOptions`, should I add some sort of filtering/conditions to `_parseQuarantineOptions`?

rob4629 commented:
> This approach doesn't seem to require a lot of changes. You need to calculate the failedThreshold value as a result of the retryCount minus passCount expression and validate if that passCount is less than retryCount.

Ok, I don't think I understand your suggestion... here's how it sounds to me (_sudo code_)

```js
const failedThreshold = passCount < retryCount ? retryCount - passCount : DEFAULT_QUARANTINE_THRESHOLD;

const failedThresholdReached = failedTimes >= failedThreshold;
....
```

```sh
testcafe -q                                // no change
testcafe -q passCount=1                    // failedCount = 2
testcafe -q retryCount=5                   // failedCount = 2
testcafe -q retryCount=10                  // failedCount = 7
testcafe -q retryCount=10,passCount=1      // failedCount = 9

miherlosev commented:
In general, you understand the explanation correctly.
Just one correction: you should validate that the passCount is less than retryCount.

Pseudocode
```js
validateArguments({ retryCount, passCount });

const failedThreshold = retryCount - passCount;

const failedThresholdReached = failedTimes >= failedThreshold;

// All correct
testcafe -q                                // no change
testcafe -q passCount=1                    // failedThreshold = 2
testcafe -q retryCount=5                   // failedThreshold = 2
testcafe -q retryCount=10                  // failedThreshold = 7
testcafe -q retryCount=10,passCount=1      // failedThreshold = 9

// Raise an error about incorrect parameters
testcafe -q retryCount=3,passCount=4
```

rob4629 commented:
Ok, I've added further validation to the `argument-parser`, plus 2 new tests to verify _passCount_ < _retryCount_.

I've also implemented the failedThreshold as per your request, however I believe we should have a minimum _failedThreshold_ of 3, rather than setting _failedThreshold_ to 1 for `-q retryCount=4` or `-q passCount=1`.

Ultimately the decision is yours, but I look forward to hearing your thoughts.

viktoria2506 commented:
I think that in the previous comment we made a mistake and `failedThreshold` should be calculated like this: `failedThreshold = retryCount - passCount + 1`. 
And also for compatibility with the current implementation, the values of the variables should be `passCount = 3 `and `retryCount = 5` by default. The remaining parameters are calculated based on these values.

rob4629 commented:
Is there anything I need to do, in order for this PR to be reviewed?

I understand if it's a case of just waiting until you have availability. I just want to make sure I haven’t forgotten to do something.

AlexKamaev commented on src/cli/argument-parser.ts:
```diff
@@ -273,6 +275,11 @@ export default class CLIArgumentParser {
     }
 
+    private async _parseQuarantineOptions (): Promise<void> {
+        if (this.opts.quarantineMode && this.opts.quarantineMode.length > 0)
```
As I understand, here the `this.opts.quarantineMode` is string. If so, we do not need to check the `length` property.

AlexKamaev commented on src/utils/get-options/quarantine.ts:
```diff
@@ -0,0 +1,39 @@
+function validateQuarantineOptions (options: Dictionary<string | number> ): void {
+    const { retryCount, passCount } = options;
+
+    if (!passCount && retryCount < 3)
```
Here we check the `passCount` variable. And if it does not exists or equal to 0 we throw the `invalidRetryCount` error. It looks there is a mistake here. I mean, that we should not throw throw `retryCount` error if we have incorrect `passCount`.

rob4629 replied:
The reason I implemented it this way is in the scenario where _retryCount_ is specified without _passCount_, then we want to ensure _retryCount_ is above the default quarantine threshold value (3).

```sh
testCafe chrome -q                // PassThreshold is 3, FailThreshold is 3
testcafe chrome -q retryCount=10  // PassThreshold is 3, FailThreshold is 8
testcafe chrome -q passCount=0    // Should fail, will add a test and validate this
testcafe chrome -q passCount=4    // Should fail, passCount > retryCount (default 3)
testcafe chrome -q retryCount=2   // Should fail, passCount (default 3) > retryCount
```

I can change the error to be something like: `passCount should be less than retryCount (retryCount value)`?

AlexKamaev replied:
Ok, I understand your idea. Thank you for explanation. What if we rewrite the code as follows:
```JS
function validateQuarantineOptions (options: Dictionary<string | number> ): void {
    const retryCount = options.retryCount || DEFAULT_TEST_RUN_THRESHOLD;
    const passCount  = options.passCount || DEFAULT_QUARANTINE_THRESHOLD;

    if (passCount > retryCount)
        throw new GeneralError(RUNTIME_ERRORS.invalidRetryCountValue, passCount);
}
```
Here I added default values for passCount and retryCount, so we do not need to check the `passCount` variable for undefined.

Could you please check this approach?
I would also prefer to import `DEFAULT_TEST_RUN_THRESHOLD` and `DEFAULT_QUARANTINE_THRESHOLD` constants. I think we can extract the Quarantine class from the `test-run-controller` file to the `quarantine` file and import these constants from the `quarantine` file.

AlexKamaev commented on src/runner/test-run-controller.ts:
```diff
@@ -24,9 +25,13 @@ interface AttemptResult {
 
     public constructor () {
         this.attempts = [];
+        this.TEST_RUN_THRESHOLD = DEFAULT_TEST_RUN_THRESHOLD;
```
Please make not constant members in camelCase.
You can make this members as properies with getter/setter since you already have special set functions for them

AlexKamaev commented on src/runner/test-run-controller.ts:
```diff
@@ -69,6 +83,13 @@ class Quarantine {
+            return this.TEST_RUN_THRESHOLD - this.PASSED_QUARANTINE_THRESHOLD + 1;
+
+        return DEFAULT_QUARANTINE_THRESHOLD;
+    }
```
you can make the new property from the `_getFailedThreshold` function as well. It's up to you

AlexKamaev commented on test/server/cli-argument-parser-test.js:
```diff
@@ -644,6 +644,51 @@ describe('CLI argument parser', function () {
             });
     });
 
+    describe('Quarantine Options', function () {
```
Please add some functional test for this feature. We already implemented some functional tests in quarantine. You can take a look on it:
https://github.com/DevExpress/testcafe/blob/e154ecfb5d40ebc92a32ceab87e82a335701d5bd/test/functional/fixtures/api/es-next/take-screenshot/testcafe-fixtures/take-screenshot.js#L58

rob4629 replied:
Thanks for the suggestion. I'm getting an error when trying to run the functional tests locally (`Error: Unable to find the browser. "ie" is not a browser alias or path to an executable file.`).

Is there a way I can trigger these without IE being installed?

AlexKamaev replied:
Do you mean tasks in gulpfile.js?
Yes, we have tasks to run tests without IE: https://github.com/DevExpress/testcafe/blob/e3c1fea1e1631777ed8a643503da50ca397b70a9/Gulpfile.js#L824

rob4629 replied:
Ahh great, I didn't see that one 🤦🏻‍♂️. Thanks. I'll push in my changes later today.

rob4629 commented:
I've added 3 tests, please let me know if you need more coverage. It also highlighted a bug I had introduced in the `_setFailedThreshold()` function... so thanks for the prompt!

rob4629 commented:
Do you need me to do more for this PR? It seems like my local master branch has become corrupted... So I'm trying to resolve that before I can investigate these failing tests

DIRECTcut commented:
@rob4629 Great PR. 

We've moved our documentation to a separate repository and it's not available for open-source contributions anymore. Please rebase your PR and remove all documentation changes (changes to `.md` files) from this PR.

I'll make sure that the documentation for this new feature makes it into the docs.

Thanks.
